2_process/src/combine_NLCD_PRMS.R

```diff
@@ -0,0 +1,64 @@
+AOI_LC_w_area <- function(NLCD_LC_df, area_att, aoi_comids_df){
+  
+  #' @description following the comid id as the main key, this function subset NLCD LC data to the area of interest, joins with nhd COMID area info and PRMS info
+  #' Output is a df with PRMS ids, COMIDs, AREA(SQKM), and NLCD LC percentage per COMID (unique key)
+  #' @param area_att df with comids and associated area (AREASQKM)
+  #' @param NLCD_LC_df selected LC dataframe structured by COMID 
+  #' @param aoi_comids_df df linking comid and prms for the area of interest, already filter to area of interest 
+  #' @example AOI_LC_w_area(NLCD_LC_df = p1_NLCD_df$NLCD_LandCover_2011, area_att = p2_nhd_area_att, drb_comids_df = p2_drb_comids)
+  
+# Filter land cover attributes to aoi comids (drb) and join with area columns
+  lc_area <- NLCD_LC_df %>%
+    filter(COMID %in% aoi_comids_df$comid) %>%
+    left_join(., area_att, by="COMID") %>%
+    mutate(COMID = as.character(COMID))
+
+# Create df with PRMS, COMID, and NLCD LC att for the aoi (drb)
+  lc_aoi <- aoi_comids_df %>%
+    left_join(., lc_area, by = c('comid' = "COMID"))
+
+  return(lc_aoi)
+
+  }
+
+###----
+# Estimate proportion of LC coverage at PRMS scale
+proportion_lc_by_prms <- function(NLCD_LC_df_w_area,
+                          catchment_att = "CAT"){
+
+#' @description This function takes the consolidated df from AOI_LC_w_area() and calculated the percentage share of each LC class at the PRMS segid scale
+#' @param NLCD_LC_df_w_area df output of AOI_LC_w_area that has LC data per COMID & PRMS id and area in (km2)
+#' @param catchment_att Specifies the catchment attribute to subset to. Must be either 'CAT', "ACC", 'TOT'. Default = "CAT"
+#' @example proportion_lc_by_prms(NLCD_LC_df_w_area = p2_LC_w_catchment_area, catchment_att = "CAT")
+
+  area_df <- NLCD_LC_df_w_area %>%
+    # multiply COMID area by the LC percentage for all the cols that start with 'CAT" or "ACC" or "TOT"
+    mutate(across(starts_with(catchment_att),~(.x/100)*AREASQKM,.names="AREA_{col}")) %>%
+    # aggregate to PRMS scale through group_by()
+    group_by(PRMS_segid) %>%
+    # summarize group by at PRMS catchment scale
+    summarize(
+      # Sum to calculate total area of the PRMS  catchment
+      across(AREASQKM, sum, .names='AREASQKM_PRMS'),
+      # Sum to calculate  LC area at the PRMS catchment scale per LC class
+      across(starts_with(paste0('AREA_',catchment_att)), sum, .names = 'PRMS_{col}'),
+      # calculate percentage coverage of specified LC class in the entire PRMS catchment
+      across(starts_with(paste0('PRMS_AREA_',catchment_att)), ~(.x/AREASQKM_PRMS),
+             .names="PERCENT_{col}"),
+      # round the new cols 
+      across(starts_with(c('PERCENT','PRMS')), round, 2), 
+    .groups="drop")
+  
+  ## Final cleaning - removing LC categories and PRMS_segis that are empty and the empty PRMS (287_1) 
+  area_df <- area_df %>%
+    filter(AREASQKM_PRMS > 0) %>%
+    select(-contains('NLCD11_12'))
```

Comment:
I see how COMIDs where `AREASQKM=0` cause an issue for at least one PRMS segment here, segid "287_1". It makes sense to handle those cases, but filtering that segment out of the data will leave us with a data frame 458 rows long (instead of 459) and I wonder if we'll ever find that confusing when working with these data down the line. Instead of filtering out this row, what if we used `across` to conditionally replace all the land cover values with NA if `AREASQKM_PRMS=0`? Maybe translating this [SO post](https://stackoverflow.com/questions/62578575/conditionally-replace-values-across-multiple-columns-based-on-string-match-in-a) would look like:

`mutate(across(c(PERCENT_AREA_CAT_first_col:PERCENT_AREA_CAT_last_col), ~ifelse(AREASQKM_PRMS==0, NA, .)))`

Do you think this makes sense, or do you think it's better to just exclude "287_1"?

Reply:
If we do not have attributes for a reach, the reach would have to be excluded from modeling. So using NA for those reaches is fine. But we know that `AERASQKM` must be greater than 0 for every reach, so I'm wondering if we need to handle these cases before computing the PRMS land cover values. Could you summarize the lengths of the comids with `AREASQKM = 0`? Maybe (length of 0 area comid) / (length of PRMS segment). I am hoping the 0 area segments are all short relative to the length of the PRMS segment, so excluding them from the land cover calculation will have a minor impact.

Reply:
> And for the 0 area PRMS segment, we could drop it from modeling. Do you know how many SC samples were assigned to that reach?

Right now, it looks like there are no SC samples that are assigned to PRMS segid "287_1":
```
> tar_load(p2_sites_w_segs)
> p2_sites_w_segs %>% filter(subsegid == "287_1")
# A tibble: 0 x 14
# ... with 14 variables: site_id <chr>, site_name <chr>, count_days_nwis <dbl>, count_days_discrete <dbl>, count_days_total <dbl>, lon <dbl>,
#   lat <dbl>, datum <chr>, org_id <chr>, data_src_combined <chr>, subsegid <chr>, offset <dbl>, segidnat <int>, geometry <GEOMETRY [°]>
```

Reply:
Here's a quick attempt at estimating the proportional length represented by the comids where `AREASQKM=0` -

```
tar_load(p2_prms_nhdv2_xwalk)
tar_load(p1_nhdv2reaches_sf)
tar_load(p1_reaches_sf)

drb_comids <- p2_prms_nhdv2_xwalk %>% 
  tidyr::separate_rows(comid_all,sep=";") %>%
  mutate(AREASQKM=NA, length_km = NA, prms_length_km = NA)

for(i in seq_along(drb_comids$PRMS_segid)){
  nhd_line <- filter(p1_nhdv2reaches_sf,COMID == drb_comids$comid_all[i])
  prms_line <- filter(p1_reaches_sf,subsegid == drb_comids$PRMS_segid[i])
  drb_comids$AREASQKM[i] <- nhd_line$AREASQKM
  drb_comids$length_km[i] <- as.numeric(st_length(nhd_line))/1000
  drb_comids$prms_length_km[i] <- as.numeric(st_length(prms_line))/1000
}

drb_comids %>% 
  mutate(comid_prop_len = length_km/prms_length_km) %>%
  filter(AREASQKM==0) %>% 
  filter(comid_prop_len > 0.05)
```

This returns 5 instances (out of the 47 `AREASQKM=0` comids that Margaux identified) where the proportional length is greater than 5%:
```
# A tibble: 5 x 7
  PRMS_segid comid_down comid_all AREASQKM length_km prms_length_km comid_prop_len
  <chr>           <int> <chr>        <dbl>     <dbl>          <dbl>          <dbl>
1 226_1         4151820 4151818          0   0.00580         0.109          0.0532
2 287_1         4188275 4188275          0   0.0480          0.0479         1.00  
3 51_1          2588373 2588371          0   0.0123          0.105          0.117 
4 878_1         4782271 4782271          0   0.0432          0.0620         0.696 
5 892_1         4782565 4782567          0   0.0143          0.0471         0.303 
```

So I wouldn't expect these comids to have a big impact on the calculated land cover proportions. It looks like there are a few SC observations that are snapped to these PRMS segments:
```
> p2_sites_w_segs %>% filter(subsegid %in% c("226_1","287_1","51_1","878_1","892_1"))
# A tibble: 4 x 14
  site_id      site_name         count_days_nwis count_days_discr~ count_days_total   lon   lat datum org_id  data_src_combin~ subsegid  offset segidnat
  <chr>        <chr>                       <dbl>             <dbl>            <dbl> <dbl> <dbl> <chr> <chr>   <chr>            <chr>      <dbl>    <int>
1 11NPSWRD_WQ~ VALLEY CREEK DS ~              NA                 3                3 -75.5  40.1 NAD83 11NPSW~ Harmonized_WQP_~ 892_1    4.01e-4     2326
2 01444800     Delaware River n~              NA               147              147 -75.1  40.8 NAD83 USGS-PA Harmonized_WQP_~ 51_1     1.15e-3       NA
3 21PA_WQX-WQ~ DELAWARE RIVER                 NA                86               86 -75.1  40.8 NAD83 21PA_W~ Harmonized_WQP_~ 51_1     1.15e-3       NA
4 31DRBCSP-DR~ Delaware River a~              NA                 7                7 -75.1  40.8 NAD83 31DRBC~ Harmonized_WQP_~ 51_1     9.54e-4       NA
# ... with 1 more variable: geometry <LINESTRING [°]>
```

Reply:
Thanks for checking that. I'm glad it's just a few PRMS reaches that are affected by this problem. 

I just realized that dropping the PRMS reach whose area is 0 may not be a good plan because it is used in the network connection tables and is an interior segment (as opposed to being on the edge or being a headwater segment). It's also used in the PRMS and streamflow prediction models that will provide input features for this model. So it's better if we can use this reach in the model. Maybe we can approximate the area for just this reach as `length_km^2`?

Reply:
@jds485  Circling back to this.

Why length_km^2? 
This approximation does put the value into the area units, however assumes that the catchment is a square around the  reach length, where the length of reach (LENGTHKM) is a side. At least that how I am seeing this.
How about assuming the cross-section of the catchment polygon as `LENGTHKM`. So equiv. to a diagonal of square shape. 
 
 in that case, the new approximated area `= 1/2 length_km^2`

Reply:
Reach length squared does not force the catchment to be shaped like a square, but I understand your reasoning here. I'm not sure which would be a better estimator of true catchment area. We could test it out with the shorter reaches in our dataset to be formal (e.g., by mean square error)

Reply:
I did not run any mean square error testing, but I've evaluated the rougher options we have to estimate the missing area. 

The difference between running  `1/2 length_km^2` and just `length_km^2` is relatively negligible in any case. 
When I apply both to get an approx area for those comids with area =0, the sums differ by 1 km2. 
```
> ## With Approximated area calc as LENGTHKM^2
> area_df_1 <-  p2_LC_w_catchment_area %>%
+   mutate(AREASQKM_approx = case_when(AREASQKM == 0 ~ LENGTHKM^2, TRUE  ~ AREASQKM))
> 
> ## With Approximated area calc as 1/2*LENGTHKM^2
> area_df_2 <-  p2_LC_w_catchment_area %>%
+   mutate(AREASQKM_approx = case_when(AREASQKM == 0 ~ 1/2*(LENGTHKM^2), TRUE  ~ AREASQKM))
> 
> sum(area_df_1$AREASQKM_approx[area_df_d$AREASQKM == 0])
[1] 1.97574
> sum(area_df_2$AREASQKM_approx[area_df_d$AREASQKM == 0])
[1] 0.98787
```
Per comid - we evidently see that `AREASQKM_approx` calc. in `area_df_2` is half that of `area_df_1`

A potential alternative could be to apply to the comid with `area == 0` the mean `AREASQKM` of comids with similar `LENGTHKM`. So, apply the mean `AREASQKM` value of comids with `AREASQKM > 0` & `LENGTHKM` ~= to `LENGTHKM` of comids `area == 0`. 

```
> #alternative 
> ## Takea average lengthkm of comid with 0 area
> avg_length_0_area <- mean(p2_LC_w_catchment_area$LENGTHKM[p2_LC_w_catchment_area$AREASQKM == 0])
> print(avg_length_0_area)
[1] 0.04240077
> 
> # extract area of other comids that are ~ the same length
> mean(p2_LC_w_catchment_area$AREASQKM[round(p2_LC_w_catchment_area$LENGTHKM,2) == round(avg_length_0_area,2)])
[1] 0.0309473
```

Now, ultimately it is prms `287_1` that turning to NA comids are aggregated to PRMS level, because that segid is made up of only one comid (`4188275` ).  For the other PRMS with comids with `area == 0`,  the `AREASQKM` of that PRMS is estimated by the other valid comid areas that it comprises. 
The `LENGTHKM` of PRMS `287_1` is `0.048`. Could we take the average area of all catchments in our basin with a LENGTHKM of 0.048 ? 

```
> len_287_1 <-  p2_LC_w_catchment_area$LENGTHKM[p2_LC_w_catchment_area$PRMS_segid == '287_1']
> p2_LC_w_catchment_area$AREASQKM[p2_LC_w_catchment_area$LENGTHKM == len_287_1]
 [1] 0.0045 0.0018 0.0018 0.0000 0.0000 0.0369 0.0018 0.0027 0.0072 0.0099 0.0000 0.0000 0.0009 0.0099 0.0018 2.1951 0.0018 0.0063 0.0090 0.0036 0.0036 0.0018 0.0000 0.0027 0.0207 0.0054
> mean(p2_LC_w_catchment_area$AREASQKM[p2_LC_w_catchment_area$LENGTHKM == len_287_1])
[1] 0.08958462

```

While I've outlined 3 options here, I left our initial option  (LENGTHKM^2) in the code for now.

Reply:
I'm in favor of either 1) assuming area ~ length^2 for any COMID with area = 0, or 2) assuming area ~ some DRB mean for very short reaches.

I'm tempted to go with option 1 just for simplicity. As far as I can tell, the longest NHD reach with zero area is 193 meters. The reach that is equivalent to PRMS segid 287_1 is 48 meters long. So these are really small and I'm guessing our decision here will have a negligible impact on the variable estimates at the PRMS-scale. Even for 287_1, for instance, there are no SC samples that currently get snapped to that segment, so I think assigning an area is less about getting an accurate estimate and more about having a placeholder value so that the segment gets retained in this data processing step. What do you think, @jds485?